From 05eaf2faa36355c93ab713a55b2540ff6ad0dec3 Mon Sep 17 00:00:00 2001
From: Michal Suchanek <msuchanek@suse.de>
Date: Sun, 11 Nov 2018 21:23:53 +0100
Subject: [PATCH] mmc: bcm2835: reset host on timeout

References: bsc#1070872
Patch-mainline: v5.0-rc1
Git-commit: f6000a4eb34e6462bc0dd39809c1bb99f9633269

The bcm2835 mmc host tends to lock up for unknown reason so reset it on
timeout. The upper mmc block layer tries retransimitting with single
blocks which tends to work out after a long wait.

This is better than giving up and leaving the machine broken for no
obvious reason.

Fixes: 660fc733bd74 ("mmc: bcm2835: Add new driver for the sdhost controller.")
Signed-off-by: Michal Suchanek <msuchanek@suse.de>
Signed-off-by: Stefan Wahren <stefan.wahren@i2se.com>
Acked-by: Eric Anholt <eric@anholt.net>
Signed-off-by: Ulf Hansson <ulf.hansson@linaro.org>
---
 drivers/mmc/host/bcm2835-sdhost.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/mmc/host/bcm2835-sdhost.c b/drivers/mmc/host/bcm2835-sdhost.c
index 29f78546ee7d..b5522806d5d2 100644
--- a/drivers/mmc/host/bcm2835-sdhost.c
+++ b/drivers/mmc/host/bcm2835-sdhost.c
@@ -336,6 +336,7 @@ static void bcm2835_sdhost_reset(struct mmc_host *mmc)
 
 	if (host->dma_chan)
 		dmaengine_terminate_all(host->dma_chan);
+	host->dma_chan = NULL;
 	tasklet_kill(&host->finish_tasklet);
 	bcm2835_sdhost_reset_internal(host);
 }
@@ -1022,6 +1023,8 @@ static void bcm2835_sdhost_timeout(unsigned long data)
 		       mmc_hostname(host->mmc));
 		bcm2835_sdhost_dumpregs(host);
 
+		bcm2835_sdhost_reset(host->mmc);
+
 		if (host->data) {
 			host->data->error = -ETIMEDOUT;
 			bcm2835_sdhost_finish_data(host);
-- 
2.20.1

